### PR TITLE
Fixes NoSuchMethodError on processing errors in pmd-compat6

### PR DESCRIPTION
**Pull request by bergander**
## Describe the PR

When a processing error occurs while using pmd-compat6 (rc4) then a NoSuchMethodError is thrown:

`[ERROR] Failed to execute goal org.apache.maven.plugins:maven-pmd-plugin:3.21.2:pmd (default-cli) on project flexpay-reflectdb: Execution default-cli of goal org.apache.maven.plugins:maven-pmd-plugin:3.21.2:pmd failed: An API incompatibility was encountered while executing org.apache.maven.plugins:maven-pmd-plugin:3.21.2:pmd: java.lang.NoSuchMethodError: 'java.lang.String net.sourceforge.pmd.Report$ProcessingError.getFile()'`

This PR implements the missing getFile() method.

## Ready?

- [ ] Added unit tests for fixed bug/feature
- [ ] Passing all unit tests
- [ ] Complete build `./mvnw clean verify` passes (checked automatically by github actions)
- [ ] Added (in-code) documentation (if needed)

### CODE REVIEW

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">No regression tested rules have been changed.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Review by adangel — CHANGES_REQUESTED** [review]
Thanks for the PR.

Could you please add an integration test (see https://github.com/pmd/pmd/tree/master/pmd-compat6/src/it ), so that we have a reproducer and a verifier, that this fixes the problem and so that it is not reintroduced in the future.

I assume, this is about printing processing errors (here https://github.com/apache/maven-pmd-plugin/blob/9ce35a43fef8f50d84782eb8a960ceb832d386ec/src/main/java/org/apache/maven/plugins/pmd/exec/PmdExecutor.java#L265 ).
One simple way to provoke a processing error (and therefore to reproduce the problem) is by creating a custom ruleset with one custom XPath-based rule, which uses an invalid xpath expression.

File: pmd-compat6/src/main/java/net/sourceforge/pmd/Report.java [hunk 1]
```diff
@@ -140,6 +140,10 @@ public FileId getFileId() {
             return file;
         }
 
+        public String getFile() {
+            return file.getAbsolutePath();
+        }
+
```
[the remark this answers is not in the document]
Comment (bergander) [reply pmd-compat6/src/main/java/net/sourceforge/pmd/Report.java]: Fixed!

**Comment by bergander** [comment]
A rule with an invalid XPath expression will only generate an error row in the log during initialization of the rule:

"Exception while initializing rule <rule name>, the rule will not be run"

To produce a processing error, the rule needs to be initialized successfully and then throw an exception during execution. So instead I resorted to writing a rule in Java which just throws an exception. And to use it I had to create a test-jar and include it as a dependency to the pmd maven plugin.

Also, ProcessingError.getFile() is only called when using the pmd:pmd goal and not when using pmd:check, so that had to be added as well.

I'm not sure this is the best solution, so feel free to suggest a better approach!

File: pmd-compat6/src/it/pmd-for-java/pom.xml [hunk 2]
```diff
@@ -21,6 +21,18 @@
                 <artifactId>maven-pmd-plugin</artifactId>
                 <version>@maven-pmd-plugin.version.for.integrationtest@</version>
                 <executions>
+                    <execution>
+                        <id>java-processing-error-pmd</id>
+                        <goals>
+                            <goal>pmd</goal>
+                        </goals>
+                        <phase>verify</phase>
+                        <configuration>
```
Comment (adangel) [inline pmd-compat6/src/it/pmd-for-java/pom.xml:30]: This configuration section is not applied due to https://issues.apache.org/jira/projects/MPMD/issues/MPMD-236 : the check goal actually calls the pmd goal, but it doesn't apply the configuration for the specific execution.

It's easier, if we keep only one execution here and add the rulesets to the main configuration (lines 31/43 below):

```xml

                    <rulesets>
                        <ruleset>/rulesets/java/maven-pmd-plugin-default.xml</ruleset>
                        <ruleset>${project.basedir}/exception_ruleset.xml</ruleset>
                    </rulesets>
```

Note: it explicitly uses the default maven-pmd-plugin ruleset plus the additional one to provoke a processing error.

File: pmd-compat6/src/it/pmd-for-java/exception_ruleset.xml [hunk 3]
```diff
@@ -0,0 +1,9 @@
+<ruleset xmlns="http://pmd.sourceforge.net/ruleset/2.0.0" name="Exception throwing ruleset">
+    <description/>
+    <rule name="ExceptionThrowingRule"
+          language="java"
+          class="net.sourceforge.pmd.ExceptionThrowingRule">
+        <description>Use this rule to produce a processing error.</description>
+    </rule>
```
Comment (adangel) [inline pmd-compat6/src/it/pmd-for-java/exception_ruleset.xml:7]: With the `error()` function of XPath, we can create a processing error at runtime as a simple XPath rule:

```suggestion
    <rule name="ExceptionThrowingRule"
          language="java"
          class="net.sourceforge.pmd.lang.rule.XPathRule">
        <description>Use this rule to produce a processing error.</description>
        <priority>3</priority>
        <properties>
            <property name="xpath">
                <value><![CDATA[
error()
            ]]></value></property>
        </properties>
    </rule>
```

**Review by adangel — COMMENTED** [review]
Thanks for the update!

> A rule with an invalid XPath expression will only generate an
> error row in the log during initialization of the rule:
>
> "Exception while initializing rule , the rule will not be run"

Oh, yes, my bad.

> To produce a processing error, the rule needs to be initialized 
> successfully and then throw an exception during execution. So 
> instead I resorted to writing a rule in Java which just throws an 
> exception. And to use it I had to create a test-jar and include it 
> as a dependency to the pmd maven plugin.

Instead of creating a java based rule, we can still write a XPath rule using the [error() function](https://www.w3.org/TR/xquery-operators/#func-error). This ends up as a processing error.
See my comment on the ruleset file.

So we don't need ExceptionThrowingRule and don't need to create a test-jar and this additional dependency.

> Also, ProcessingError.getFile() is only called when using the
> pmd:pmd goal and not when using pmd:check, so that had to be added 
> as well.

That's due to [MPMD-236](https://issues.apache.org/jira/projects/MPMD/issues/MPMD-236) - pmd:check is executing pmd:pmd, but doesn't apply the configuration from the specific execution. See my comment for the pom.xml.

Since we now have a processing error, we need to make sure, that we don't stop maven-pmd-plugin - therefore we need to "skipPmdError":

```
<skipPmdError>true</skipPmdError> <!-- we want to capture processing errors -->
```

Otherwise no report we be created and the verification will fail.

Also, we should verify, that the processing error is indeed ending up in the report (in target/pmd.xml). Please add the following lines in `verify.bsh` (at line 36):

```
if (!pmdXml.contains("<error filename=\"" + mainFile.getAbsolutePath()) || !pmdXml.contains(mainFile + "\" msg=\"PmdXPathException")) {
    throw new RuntimeException("Processing error has not been reported");
}
```

> I'm not sure this is the best solution, so feel free to suggest a better approach!

I'd prefer to use the XPath-based approach, so that we don't need to create this extra test rule class.

Note: the failing CI build is unrelated to your change, it looks like intermittent NPE (#4757)

**Review by adangel — APPROVED** [review]
Thanks!

**Pull request merged**